clippy_lints/src/assigning_clones.rs

```diff
@@ -0,0 +1,204 @@
+    if !(implements_trait(cx, clone_result_type, op.expected_trait(cx)?, &[])) {
+        return None;
+    }
+
+    // If we're assigning to a dereferenced reference, then we know the place is already valid.
+    // On the other hand, if the place is a variable or a Box, it might be uninitialized,
+    // in which case the suggestion might be wrong.
+    // TODO: Actually ask whether the place is uninitialized at this point, instead of
+    // guessing based on the syntax and type.
+    let ExprKind::Unary(hir::UnOp::Deref, derefed_target_expr) = assign_target.kind
+        else { return None };
+    if !cx.typeck_results().expr_ty(derefed_target_expr).is_ref() {
+        return None;
+    }
+
+    Some(())
+}
+
+fn suggest<'tcx>(
+    cx: &LateContext<'tcx>,
+    op: Op,
+    assign_expr: &hir::Expr<'tcx>,
+    assign_target: &hir::Expr<'tcx>,
+    clone_receiver: &hir::Expr<'tcx>,
+) {
+    span_lint_and_then(cx, ASSIGNING_CLONES, assign_expr.span, op.message(), |diag| {
+        // TODO: Make this MachineApplicable once we are more certain that the method being called
+        // is what we think it is.
+        let mut applicability = Applicability::MaybeIncorrect;
+
+        diag.span_suggestion(
+            assign_expr.span,
+            op.suggestion_msg(),
+            op.suggested_replacement(cx, assign_target, clone_receiver, &mut applicability),
+            applicability,
+        );
+    });
+}
+
+impl Op {
+    fn expected_trait(self, cx: &LateContext<'_>) -> Option<hir::def_id::DefId> {
+        match self {
+            Op::Clone => cx.tcx.lang_items().clone_trait(),
+            Op::ToOwned => cx.tcx.get_diagnostic_item(sym::ToOwned),
+        }
+    }
+
+    fn message(self) -> &'static str {
+        // TODO: Use the receiver type to say "is" instead of "may be" for types which
+        // are known to have optimizations (e.g. `String`).
+        match self {
+            Op::Clone => "assigning the result of `Clone::clone()` may be inefficient",
+            Op::ToOwned => "assigning the result of `ToOwned::to_owned()` may be inefficient",
+        }
+    }
+
+    fn suggestion_msg(self) -> &'static str {
+        match self {
+            Op::Clone => "use `clone_from()`",
+            Op::ToOwned => "use `clone_into()`",
+        }
+    }
+
+    fn suggested_replacement<'tcx>(
+        self,
+        cx: &LateContext<'tcx>,
+        assign_target: &hir::Expr<'tcx>,
+        clone_receiver: &hir::Expr<'tcx>,
+        applicability: &mut Applicability,
+    ) -> String {
+        match self {
+            Op::Clone => {
+                // The assignment LHS, which will become the receiver of the `.clone_from()` call,
+                // should lose one level of dereference operator since autoref takes care of that.
+                let target_sugg = if let ExprKind::Unary(hir::UnOp::Deref, ref_expr) = assign_target.kind {
+                    Sugg::hir_with_applicability(cx, ref_expr, "_", applicability)
```

Comment:
Nice usage of `hir_with_applicability` :+1: 

You can resolve this directly :)